src/hddtemp.c

```diff
@@ -247,55 +275,46 @@ static void hddtemp_submit (char *type_instance, double value)
 
 static int hddtemp_read (void)
 {
-	char buf[1024];
-	char *fields[128];
+	char *buf;
 	char *ptr;
 	char *saveptr;
-	int num_fields;
-	int num_disks;
-	int i;
+	char *name;
+	char *model;
+	char *temperature;
+	char *mode;
 
 	/* get data from daemon */
-	if (hddtemp_query_daemon (buf, sizeof (buf)) < 0)
+	buf = hddtemp_query_daemon ();
+	if (buf == NULL)
 		return (-1);
 
 	/* NB: strtok_r will eat up "||" and leading "|"'s */
-	num_fields = 0;
 	ptr = buf;
 	saveptr = NULL;
-	while ((fields[num_fields] = strtok_r (ptr, "|", &saveptr)) != NULL)
+	while ((name = strtok_r (ptr, "|", &saveptr)) != NULL &&
+	       (model = strtok_r (NULL, "|", &saveptr)) != NULL &&
+	       (temperature = strtok_r (NULL, "|", &saveptr)) != NULL &&
+	       (mode = strtok_r (NULL, "|", &saveptr)) != NULL)
```

Comment:
Since you're touching this anyway, it should use strsplit() from common.h instead.